### PR TITLE
dataflow: Update taint-mode docs with field sensitivity

### PR DESCRIPTION
**Pull request by IagoAbal**
Closes PA-1996

# Thanks for improving Semgrep Docs 😀

### Please ensure

- [x] A subject matter expert (SME) reviews the content
- [x] A technical writer reviews the content or PR
- [x] This change has no security implications or else you have pinged the security team
- [x] Redirects are added if the PR changes page URLs

### CODE REVIEW

**Comment by adamkvitek** [comment]
@IagoAbal thank you! I will check this PR asap (probably tomorrow).

File: docs/writing-rules/data-flow/data-flow-overview.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ Semgrep provides no user-friendly way of specifying a new data-flow analysis. Pl
 Semgrep strives for simplicity and delivers a lightweight, and fast static analysis. In addition to being intra-procedural, here are some other trade-offs:
 
 - No path sensitivity: All _potential_ execution paths are considered, despite that some may not be feasible.
-- No pointer or shape analysis: _Aliasing_ that happens in non-trivial ways may not be detected, such as through arrays or pointers. Individual elements in arrays or other data structures are not tracked. There is also no proper field sensitivity at present, but this may be developed in the future.
+- No pointer or shape analysis: _Aliasing_ that happens in non-trivial ways may not be detected, such as through arrays or pointers. Individual elements in arrays or other data structures are not tracked. There is limited field-sensitivity support in the taint engine, but none for constant propagation.
```
Comment (adamkvitek) [inline docs/writing-rules/data-flow/data-flow-overview.md:30]:
```suggestion
- No pointer or shape analysis: _Aliasing_ that happens in non-trivial ways may not be detected, such as through arrays or pointers. Individual elements in arrays or other data structures are not tracked. The taint engine supports limited field sensitivity, but not yet for constant propagation.
```

File: docs/writing-rules/data-flow/taint-mode.md [hunk 2]
```diff
@@ -36,6 +36,17 @@ You can find more examples of taint rules in the [Semrep Registry](https://semgr
...
+-----------------
+
+The taint engine has basic support for field sensitivity:
+
+- It can track `x.a.b` as tainted, while `x` and `x.a` as not tainted. If `x.a.b` is tainted, any extension of `x.a.b` (such as `x.a.b.c`) will also be considered tainted.
+- If `x.a` is tainted, and later `x.a.b` is sanitized, the engine knows that `x.a.b` is not tainted but `x.a` or `x.a.c` are still tainted.
+- Taint tracking is _NOT_ index sensitive, if `x.a[i]` gets tainted, we consider `x.a` itself to be tainted. Crucially, if `x.a[i]` is sanitized, then `x.a` also gets sanitized.
+
+Note that we track taint _per variable_ and not _per object in memory_, there is no alias analysis.
+
```
Comment (adamkvitek) [inline docs/writing-rules/data-flow/taint-mode.md:49]:
```suggestion
Field sensitivity
-----------------

The taint engine provides basic field sensitivity support:

- Track `x.a.b` as tainted, while you can specify that `x` and `x.a` are not tainted. If `x.a.b` is tainted, any extension of `x.a.b` (such as `x.a.b.c`) is also considered tainted.
- If `x.a` is tainted, and later `x.a.b` is sanitized, the engine detects that `x.a.b` is not tainted but `x.a` or `x.a.c` are still tainted.
- Taint tracking is **not** index sensitive, if `x.a[i]` is tainted, Semgrep considers `x.a` itself also as tainted. However, if `x.a[i]` is sanitized, then `x.a` is also sanitized.

:::note
The taint engine does not track taint **per variable** and not **per object in memory**, there is no alias analysis.
:::
```

**Review by adamkvitek — APPROVED** [review]
@IagoAbal thank you, please apply my suggestions. Looking at: https://github.com/returntocorp/semgrep/commit/756408aafebb849d89993a067f9c6f8d834df8a4 there is no need to somehow enable field sensitivity in taint rules, it works out of the box?

**Comment by IagoAbal** [comment]
> there is no need to somehow enable field sensitivity in taint rules, it works out of the box?

Correct!

**Pull request merged**